### PR TITLE
Adds support for arc-enabled virtual machines in HCI setup

### PR DESCRIPTION
**Pull request by agadekar**
- Adds support for arc-enabled virtual machines in Azure Stack HCI setup.
Users have the option to enable creation of Arc-enabled virtual machines using Vmfleet and run workloads on the fleet.
Documentation to use this feature added under Frameworks/VMFleet/ArcVM-Setup.md

- Adds a new command Set-FleetRunProfileScript which generates profile.xml and sets run profile template script in SMB directory.
This command will is useful especially for free runs using already available profiles (Peak, SQL, VDI etc) without having to opt into the stepped runs.

Sample command
```
 $profile = Get-FleetProfileXml -Name "VDI"
 Set-FleetRunProfileScript  -ProfileXml $profile
 ```

### CODE REVIEW

File: Frameworks/VMFleet/VMFleet.psm1 [hunk 1]
```diff
@@ -8665,4 +9136,152 @@ function SetCacheBehavior
             Remove-PSSession -Session $s
         }
     }
+}
+
+function Get-Salt {
+    $salt = ([char]'a'..[char]'z' + [char]'A'..[char]'Z' + [char]'0'..[char]'9' | Get-Random -Count 4 |% { [char]$_ }) -join ''
+    Write-Host "Generated Salt: $salt"
```
Comment (dl2n) [inline Frameworks/VMFleet/VMFleet.psm1:9143]: does this ever need to be output in normal practice? consider verbose if you think its really needed.

File: Frameworks/VMFleet/VMFleet.psm1 [hunk 2]
```diff
@@ -8665,4 +9136,152 @@ function SetCacheBehavior
...
+    param(
+        [Parameter()]
+        [string]
+        $Cluster = "."
+    )
+
+    $controlPath = Get-FleetPath -PathType Control $Cluster
+    $f = Join-Path $controlPath "arc.json"
+    if (-not (Test-Path $f)) {
+        Write-Host "arc.json does not exist. Current Arc mode: Disabled."
```
Comment (dl2n) [inline Frameworks/VMFleet/VMFleet.psm1:9158]: can we put this under verbose? value understood while doing initial development.

File: Frameworks/VMFleet/VMFleet.psm1 [hunk 3]
```diff
@@ -1273,6 +1273,162 @@ $CommonFunc = {
...
+        $location = (Get-AzureStackHCI | select Region).Region
+        $isLoggedIn = az login -u $azUser -p $azPassword
+        if($isLoggedIn -eq $null){
+           throw "Azure login failed with error - $isLoggedIn"
+        }
+        az account set -s $subscription
+        try{
+           $groupResult = az group show --name $resourceGroup | ConvertFrom-Json
+            if ($groupResult.properties.provisioningState -eq "Succeeded") {
+                Write-Host "Resource group $resourceGroup already exists"
```
Comment (dl2n) [inline Frameworks/VMFleet/VMFleet.psm1:1290]: consider LogOutput so it comes with a timestamp?

File: Frameworks/VMFleet/VMFleet.psm1 [hunk 4]
```diff
@@ -1273,6 +1273,162 @@ $CommonFunc = {
...
+            }
+            else {
+                 Write-Host "Creating new Resource group $resourceGroup ..."
```
Comment (dl2n) [inline Frameworks/VMFleet/VMFleet.psm1:1293]: consider LogOutput so it comes with a timestamp?

File: Frameworks/VMFleet/VMFleet.psm1 [hunk 5]
```diff
@@ -2263,259 +2440,335 @@ function New-Fleet
...
+                                                   $Stoploop = $true
+                                               }
+                                           }
+                                           catch {
+                                               if ($Retrycount -gt 10) {
+                                                   LogOutput "Could not get VM $vmname after 10 retries using Get-ClusterResource"
+                                                   $Stoploop = $true
+                                               }
+                                               else {
+                                                   Write-Host "Could not get VM $vmname at $RetryCount retry count. Retrying in 6 minutes..."
```
Comment (dl2n) [inline Frameworks/VMFleet/VMFleet.psm1:2684]: consider LogOutput so it comes with a timestamp? note the other branch of the if immediately above

File: Frameworks/VMFleet/VMFleet.psm1 [hunk 6]
```diff
@@ -8665,4 +9136,152 @@ function SetCacheBehavior
...
+        [switch]
+        $ResetSalt,
+
+        [Parameter()]
+        [string]
+        $Cluster = "."
+    )
+
+    if($PSBoundParameters.Count -eq 0){
+        Write-Host "Set atleast one property."
```
Comment (dl2n) [inline Frameworks/VMFleet/VMFleet.psm1:9201]: "at least" and this seems like an error condition - Write-Error?

Comment (dl2n) [reply Frameworks/VMFleet/VMFleet.psm1:9201]: Note that Test-ArcConfig will do that on error in the normal flow, below.

File: Frameworks/VMFleet/VMFleet.psm1 [hunk 7]
```diff
@@ -8665,4 +9136,152 @@ function SetCacheBehavior
...
+            }
+        }
+        if($retry -eq 4) {
+            throw "Unable to create a unique salt with 3 retries. Pick another resource group or run Set-ArcConfig again."
+        }
+    }
+
+    $arcConfig | ConvertTo-Json | Set-Content -Path $f -Force:$true
+
+    Write-Host "Configuration updated in $f"
```
Comment (dl2n) [inline Frameworks/VMFleet/VMFleet.psm1:9266]: Is this needed? Normal assumption would be that no error/throw is success.

Comment (agadekar) [reply Frameworks/VMFleet/VMFleet.psm1:9266]: Intention behind adding this was to make user aware of the path where this file is stored.

File: Frameworks/VMFleet/VMFleet.psm1 (hunk 6 above)
Comment (agadekar) [reply Frameworks/VMFleet/VMFleet.psm1:9201]: this is added as a fail fast mechanism as it does not make sense to move forward beyond this point if no parameter is provided by user. Test-ArcConfig helps validate if all mandatory parameters are available to vmfleet by checking existing arc.json if any and current user provided parameters.

File: Frameworks/VMFleet/VMFleet.psm1 (hunk 7 above)
Comment (hfu949) [reply Frameworks/VMFleet/VMFleet.psm1:9266]: Hi, just my 2 cents (as I tested it last night), knowing the path to the config file is good to know.

**Review by dl2n — APPROVED** [review]

**Pull request merged**